packages/hydrogen/src/customer/discovery.ts

```diff
@@ -0,0 +1,143 @@
+  return domain.toLowerCase();
+}
+
+function isValidCache(cacheEntry: {timestamp: number} | undefined): boolean {
+  return Boolean(
+    cacheEntry && Date.now() - cacheEntry.timestamp < DISCOVERY_CACHE_TTL,
+  );
+}
+
+async function fetchWithTimeout(
+  url: string,
+  timeoutMs: number = DISCOVERY_TIMEOUT,
+): Promise<Response> {
+  const controller = new AbortController();
+  const timeoutId = setTimeout(() => controller.abort(), timeoutMs);
+
+  try {
+    const response = await fetch(url, {
+      signal: controller.signal,
+      headers: {
+        Accept: 'application/json',
+        'User-Agent': USER_AGENT,
+      },
+    });
+    clearTimeout(timeoutId);
+
+    return response;
+  } catch (error) {
+    clearTimeout(timeoutId);
+    throw error;
+  }
+}
+
+export async function discoverOpenIdConfiguration(storefrontDomain: string): Promise<OpenIdConfiguration> {
+  const cacheKey = getCacheKey(storefrontDomain);
+  const cached = discoveryCache.get(cacheKey)?.openid;
+
+  if (isValidCache(cached)) {
+    return cached!.config;
+  }
+
+  const discoveryUrl = `https://${storefrontDomain}/.well-known/openid-configuration`;
+
+  const response = await fetchWithTimeout(discoveryUrl);
+  const config = (await response.json()) as OpenIdConfiguration;
+
+  const existingCache = discoveryCache.get(cacheKey) || {};
+  discoveryCache.set(cacheKey, {
+    ...existingCache,
+    openid: {
+      config,
+      timestamp: Date.now(),
+    },
+  });
+
+  return config;
+}
+
+export async function discoverCustomerAccountApi(storefrontDomain: string): Promise<CustomerAccountApiConfiguration> {
+  const cacheKey = getCacheKey(storefrontDomain);
+  const cached = discoveryCache.get(cacheKey)?.customerApi;
+
+  if (isValidCache(cached)) {
+    return cached!.config;
+  }
+  const discoveryUrl = `https://${storefrontDomain}/.well-known/customer-account-api`;
+  const response = await fetchWithTimeout(discoveryUrl);
+  const config = (await response.json()) as CustomerAccountApiConfiguration;
+
+  const existingCache = discoveryCache.get(cacheKey) || {};
+  discoveryCache.set(cacheKey, {
+    ...existingCache,
+    customerApi: {
+      config,
+      timestamp: Date.now(),
+    },
+  });
+
+  return config;
+}
+
+export function clearDiscoveryCache(storefrontDomain?: string): void {
+  if (storefrontDomain) {
+    discoveryCache.delete(getCacheKey(storefrontDomain));
+  } else {
+    discoveryCache.clear();
+  }
+}
+
+export async function discoverCustomerAccountEndpoints(storefrontDomain: string): Promise<DiscoveredEndpoints> {
+  const [openidConfig, customerApiConfig] = await Promise.all([
+    discoverOpenIdConfiguration(storefrontDomain),
+    discoverCustomerAccountApi(storefrontDomain),
+  ]);
+
+  const endpoints = {
+    graphqlApiUrl: customerApiConfig.graphql_api,
+    authorizationUrl: openidConfig.authorization_endpoint,
+    tokenUrl: openidConfig.token_endpoint,
+    logoutUrl: openidConfig.end_session_endpoint,
+  };
```

Comment:
@kdaviduik @frandiox I would like your thoughts on this, most specifically of what to do in case of if any request here fails. These endpoints should be very stable and not change very often, but at the same time, if something happens and these endpoints can't be fetched, what should the final user see?